### PR TITLE
Update python version to 3.11 for CUDA and ROCm images

### PR DESCRIPTION
**Pull request by oksanabaza**
## Description

[RHOAIENG-11843](https://issues.redhat.com/browse/RHOAIENG-11843)

Updating python version to 3.11 for Ray CUDA and Ray ROCm images

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by astefanutti** [comment]
@oksanabaza this requires #242 first so you can change the `output-image` and `additional-tag` values accordingly.

**Comment by astefanutti** [comment]
@oksanabaza I think we can switch to Python 3.11. Also the tags have to been changed accordingly here: https://github.com/opendatahub-io/distributed-workloads/blob/90f9e75dbd039698c6375b5e00d4a9dd7f1129c3/.tekton/ray-cuda-push.yaml#L26-L29

**Comment by oksanabaza** [comment]
> @oksanabaza I think we can switch to Python 3.11. Also the tags have to been changed accordingly here:
> 
> https://github.com/opendatahub-io/distributed-workloads/blob/90f9e75dbd039698c6375b5e00d4a9dd7f1129c3/.tekton/ray-cuda-push.yaml#L26-L29

do we want to keep the tag with py39 here 2.35.0-py39-cu121 or should we change it into 2.35.0-py311-cu121?

**Comment by astefanutti** [comment]
> > @oksanabaza I think we can switch to Python 3.11. Also the tags have to been changed accordingly here:
> > https://github.com/opendatahub-io/distributed-workloads/blob/90f9e75dbd039698c6375b5e00d4a9dd7f1129c3/.tekton/ray-cuda-push.yaml#L26-L29
> 
> I see that these changes are from upstream Konflux. However, since we are considering moving to Python 3.11, should we keep the tag with py39 here `2.35.0-py39-cu121`, or should we update it to `2.35.0-py311-cu121` to reflect the new Python version?

@oksanabaza you're right, the tags have to be changed to `2.35.0-py311-cu121`  and `2.35.0-py311-rocm61` as part of that PR.

**Comment by astefanutti** [comment]
@oksanabaza have you updated the tags in the Konflux pipelines? I can't see the update :)

**Comment by oksanabaza** [comment]
@astefanutti going to updated it now, thanks

**Comment by astefanutti** [comment]
@oksanabaza you don't need to create a new PR. You can update the tags in that PR.

**Comment by oksanabaza** [comment]
@astefanutti I have updated all labels, could you please check it again?

File: .tekton/ray-cuda-push.yaml [hunk 1]
```diff
@@ -24,9 +24,9 @@ spec:
   - name: revision
     value: '{{revision}}'
   - name: output-image
-    value: quay.io/modh/ray:2.35.0-py39-cu121
+    value: quay.io/modh/ray:2.35.0-py311-cu121
   - name: additional-tag
-    value: 2.35.0-py39-cu121-{{revision}}
+    value: 2.35.0-py311-cu121-{{revision}}
```
Comment (astefanutti) [inline .tekton/ray-cuda-push.yaml:29]: I think this needs to be updated as well in the pipeline for the ROCm image.

File: README.md [hunk 2]
```diff
@@ -22,7 +22,7 @@
 * `CODEFLARE_TEST_TIMEOUT_LONG` - Timeout duration for long tasks
 * `CODEFLARE_TEST_RAY_IMAGE` (Optional) - Ray image used for raycluster configuration 
 
-    NOTE: `quay.io/modh/ray:2.35.0-py39-cu121` is the default image used for creating a RayCluster resource. If you have your own custom ray image which suits your purposes, specify it in `CODEFLARE_TEST_RAY_IMAGE` environment variable.
+    NOTE: `quay.io/modh/ray:2.35.0-py311-cu121` is the default image used for creating a RayCluster resource. If you have your own custom ray image which suits your purposes, specify it in `CODEFLARE_TEST_RAY_IMAGE` environment variable.
```
Comment (astefanutti) [inline README.md:25]: We may want to default to that new image separately to coordinate with the SDK. @sutaakar @Fiona-Waters WDYT?

**Review by astefanutti — COMMENTED** [review]
I think the README files need to be updated as well.

File: README.md (hunk 2 above)
Comment (Fiona-Waters) [reply README.md:25]: Yes we should coordinate updating this with updating the sdk. Can we create the PR in the sdk or we should ask WO team to do it?

File: .tekton/ray-cuda-push.yaml (hunk 1 above)
Comment (oksanabaza) [reply .tekton/ray-cuda-push.yaml:29]: thank you, I updated the label for the ROCm image as well

File: README.md (hunk 2 above)
Comment (astefanutti) [reply README.md:25]: Thanks, I think that's OK if we create a PR ourselves.

File: examples/hpo-raytune/notebook/raytune-oai-MR-gRPC-demo.ipynb [hunk 3]
```diff
@@ -53,7 +53,7 @@
     "    worker_memory_limits=4,\n",
     "    head_extended_resource_requests={'nvidia.com/gpu':0},\n",
     "    worker_extended_resource_requests={'nvidia.com/gpu':0},\n",
-    "    image=\"quay.io/modh/ray:2.35.0-py39-cu121\"\n",
+    "    image=\"quay.io/modh/ray:2.35.0-py311-cu121\"\n",
```
Comment (astefanutti) [inline examples/hpo-raytune/notebook/raytune-oai-MR-gRPC-demo.ipynb:56]: I'll do these updates in a follow up PR so that the image is built and available in Quay.

File: README.md (hunk 2 above)
Comment (oksanabaza) [reply README.md:25]: I can do that if you don't mind

**Review by astefanutti — COMMENTED** [review]
@oksanabaza could you move the changes in the example notebooks in a follow up PR, so we have the new images built and then we can work on updating the examples?

File: .tekton/ray-rocm-push.yaml [hunk 4]
```diff
@@ -24,7 +24,7 @@ spec:
   - name: revision
     value: '{{revision}}'
   - name: output-image
-    value: quay.io/modh/ray:2.35.0-py39-rocm61
+    value: quay.io/modh/ray:2.35.0-py311-rocm61
   - name: additional-tag
     value: 2.35.0-py39-rocm61-{{revision}}
```
Comment (astefanutti) [inline .tekton/ray-rocm-push.yaml:29]:
```suggestion
    value: 2.35.0-py311-rocm61-{{revision}}
```

File: examples/hpo-raytune/notebook/raytune-oai-MR-gRPC-demo.ipynb (hunk 3 above)
Comment (astefanutti) [inline examples/hpo-raytune/notebook/raytune-oai-MR-gRPC-demo.ipynb:56]: Can this be moved in a separate PR?

File: examples/hpo-raytune/notebook/raytune-oai-demo-mlmd.ipynb [hunk 5]
```diff
@@ -54,7 +54,7 @@
     "    min_memory=4,\n",
     "    max_memory=4,\n",
     "    num_gpus=0,\n",
-    "    image=\"quay.io/modh/ray:2.35.0-py39-cu121\",                                           \n",
+    "    image=\"quay.io/modh/ray:2.35.0-py311-cu121\",                                           \n",
```
Comment (astefanutti) [inline examples/hpo-raytune/notebook/raytune-oai-demo-mlmd.ipynb:57]: Can this be moved in a separate PR?

File: examples/hpo-raytune/notebook/raytune-oai-demo.ipynb [hunk 6]
```diff
@@ -55,7 +55,7 @@
     "    min_memory=4,\n",
     "    max_memory=4,\n",
     "    num_gpus=0,\n",
-    "    image=\"quay.io/modh/ray:2.35.0-py39-cu121\"\n",
+    "    image=\"quay.io/modh/ray:2.35.0-py311-cu121\"\n",
```
Comment (astefanutti) [inline examples/hpo-raytune/notebook/raytune-oai-demo.ipynb:58]: Can this be moved in a separate PR?

File: examples/ray-finetune-llm-deepspeed/README.md [hunk 7]
```diff
@@ -125,7 +125,7 @@ This example has been validated on the following configurations:
         head_memory=96,
         head_extended_resource_requests={'amd.com/gpu':1},
         worker_extended_resource_requests={'amd.com/gpu':1},
-        image="quay.io/rhoai/ray:2.35.0-py39-rocm61-torch24-fa26",
+        image="quay.io/rhoai/2.35.0-py311-rocm61-torch24-fa26",
```
Comment (astefanutti) [inline examples/ray-finetune-llm-deepspeed/README.md:128]: Can this be moved in a separate PR?

File: examples/ray-finetune-llm-deepspeed/ray_finetune_llm_deepspeed.ipynb [hunk 8]
```diff
@@ -68,11 +68,11 @@
     "    worker_memory_limits=256,\n",
     "    head_memory=128,\n",
     "    # Use the following parameters with NVIDIA GPUs\n",
-    "    image=\"quay.io/rhoai/ray:2.35.0-py39-cu121-torch24-fa26\",\n",
+    "    image=\"quay.io/rhoai/ray:2.35.0-py311-cu121-torch24-fa26\",\n",
```
Comment (astefanutti) [inline examples/ray-finetune-llm-deepspeed/ray_finetune_llm_deepspeed.ipynb:71]: Can this be moved in a separate PR?

File: examples/ray-finetune-llm-deepspeed/ray_finetune_llm_deepspeed.ipynb [hunk 9]
```diff
@@ -68,11 +68,11 @@
...
     "    head_extended_resource_requests={'nvidia.com/gpu':1},\n",
     "    worker_extended_resource_requests={'nvidia.com/gpu':1},\n",
     "    # Or replace them with these parameters for AMD GPUs\n",
-    "    # image=\"quay.io/rhoai/ray:2.35.0-py39-rocm61-torch24-fa26\",\n",
+    "    # image=\"quay.io/rhoai/2.35.0-py311-rocm61-torch24-fa26\",\n",
```
Comment (astefanutti) [inline examples/ray-finetune-llm-deepspeed/ray_finetune_llm_deepspeed.ipynb:75]: Can this be moved in a separate PR?

File: examples/hpo-raytune/notebook/raytune-oai-MR-gRPC-demo.ipynb (hunk 3 above)
Comment (oksanabaza) [reply examples/hpo-raytune/notebook/raytune-oai-MR-gRPC-demo.ipynb:56]: yes, thanks, I will move the following changes into a separate PR

File: examples/ray-finetune-llm-deepspeed/ray_finetune_llm_deepspeed.ipynb [hunk 10]
```diff
@@ -72,7 +72,7 @@
     "    head_extended_resource_requests={'nvidia.com/gpu':1},\n",
     "    worker_extended_resource_requests={'nvidia.com/gpu':1},\n",
     "    # Or replace them with these parameters for AMD GPUs\n",
-    "    # image=\"quay.io/rhoai/ray:2.35.0-py39-rocm61-torch24-fa26\",\n",
```
Comment (astefanutti) [inline examples/ray-finetune-llm-deepspeed/ray_finetune_llm_deepspeed.ipynb:75]: This can be moved as well for a follow up.

**Comment by oksanabaza** [comment]
@astefanutti could you please take a look one more time? I have also separated the Python version and image tag into two separate arguments in the Dockerfile

**Comment by astefanutti** [comment]
> @astefanutti I have updated all labels, could you please check it again?

@oksanabaza there are two leftovers in the examples, and then it'll be good to go.

File: examples/ray-finetune-llm-deepspeed/README.md [hunk 11]
```diff
@@ -125,7 +125,7 @@ This example has been validated on the following configurations:
         head_memory=96,
         head_extended_resource_requests={'amd.com/gpu':1},
         worker_extended_resource_requests={'amd.com/gpu':1},
-        image="quay.io/rhoai/ray:2.35.0-py39-rocm61-torch24-fa26",
+        image="quay.io/rhoai/2.35.0-py39-rocm61-torch24-fa26",
```
Comment (astefanutti) [inline examples/ray-finetune-llm-deepspeed/README.md:128]: This change should be reverted :).
